packages/SwingSet/src/vats/comms/dispatch.js

```diff
@@ -73,64 +72,20 @@ export function buildCommsDispatch(syscall) {
   }
 
-  function notifyOnePromise(promiseID, rejected, data, doNotSubscribeSet) {
-    insistCapData(data);
-    // console.debug(`comms.notifyOnePromise(${promiseID}, ${rejected}, ${data})`);
-    // dumpState(state);
-
-    // I *think* we should never get here for local promises, since the
-    // controller only does sendOnly. But if we change that, we need to catch
-    // locally-generated promises and deal with them.
-    // if (promiseID in localPromises) {
-    //  resolveLocal(promiseID, { type: 'data', data });
-    // }
-
-    // todo: if we previously held resolution authority for this promise, then
-    // transferred it to some local vat, we'll have subscribed to the kernel
-    // to hear about it. If we then get the authority back again, we no longer
-    // want to hear about its resolution (since we're the ones doing the
-    // resolving), but the kernel still thinks of us as subscribing, so we'll
-    // get a bogus dispatch.notify. Currently we throw an error, which is
-    // currently ignored but might prompt a vat shutdown in the future.
-
-    // TODO: The following goofiness, namely taking apart the capdata object
-    // and looking at it to see if it's a single presence reference and then
-    // treating it in a special way if it is, is a consequence of an impedance
-    // mismatch that has grown up between the comms protocol and the
-    // evolutionary path that the kernel/vat interface has taken.  In the
-    // future we should clean this up and unify presences and data in the
-    // comms protocol the way the kernel/vat interface has.
-    const unser = JSON.parse(data.body);
-    let resolution;
-    if (rejected) {
-      resolution = harden({ type: 'reject', data });
-    } else if (
-      Object(unser) === unser &&
-      QCLASS in unser &&
-      unser[QCLASS] === 'slot'
-    ) {
-      const slot = data.slots[unser.index];
-      insistVatType('object', slot);
-      resolution = harden({ type: 'object', slot });
-    } else {
-      resolution = harden({ type: 'data', data });
-    }
-    resolveFromKernel(promiseID, resolution, doNotSubscribeSet);
-    // XXX question: do we need to call retirePromiseIDIfEasy (or some special
-    // comms vat version of it) here?
-  }
-
   function notify(resolutions) {
     const willBeResolved = new Set();
     for (const resolution of resolutions) {
-      const [vpid] = resolution;
+      const [vpid, value] = resolution;
       willBeResolved.add(vpid);
+      assert(typeof value === 'object');
+      insistCapData(value.data);
+      // console.debug(`comms.notify(${vpid}, ${value})`);
+      // dumpState(state);
     }
+    resolveFromKernel(resolutions, willBeResolved);
 
-    for (const resolution of resolutions) {
-      const [vpid, vp] = resolution;
-      notifyOnePromise(vpid, vp.rejected, vp.data, willBeResolved);
-    }
+    // XXX question: do we need to call retirePromiseIDIfEasy (or some special
+    // comms vat version of it) here?
```

Comment:
not here, because it needs to be done after the kernel-facing vrefs have been translated into comms-vat-local lrefs

in fact it might make sense to have `resolveFromKernel` be responsible for accumulating the `willBeResolved` list, and `notify()` becomes a no-op. hm.